contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,570 @@
+
+    Cover storage currentCover = covers[coverId];
+    require(currentCover.amount > amountReduction, "Cover: amountReduction > cover.amount");
+
+    // clone the existing cover
+    Cover memory newCover = covers[coverId];
+
+    // clone existing staking pools
+    StakingPool[] memory newStakingPools = stakingPoolsForCover[coverId];
+
+    uint newTotalCoverAmount = newCover.amount - amountReduction;
+
+    uint newCoverId = covers.length;
+
+    // reduce amount
+    for (uint i = 0; i < newStakingPools.length; i++) {
+      IStakingPool stakingPool = IStakingPool(newStakingPools[i].poolAddress);
+
+      // reduce the amount per pool proportionately to the overall reduction
+      uint newCoverAmount = newStakingPools[i].coverAmount * newTotalCoverAmount / newCover.amount;
+      stakingPool.reduceAmount(
+        newCover.productId,
+        newCover.period,
+        newCover.start,
+        REWARD_BPS * newStakingPools[i].premiumInNXM / BASIS_PRECISION,
+        newCoverAmount,
+        REWARD_BPS * (newStakingPools[i].premiumInNXM * newTotalCoverAmount / newCover.amount) / BASIS_PRECISION,
+        newStakingPools[i].coverAmount
+      );
+
+      // TODO: fix this. it should be proportional to the remaining period as well
+      newStakingPools[i].premiumInNXM =
+      uint96(uint(newStakingPools[i].premiumInNXM) * newTotalCoverAmount / newCover.amount);
+      newStakingPools[i].coverAmount = uint96(newCoverAmount);
+
+      // write the new staking pool with modified parameters
+      stakingPoolsForCover[newCoverId].push(newStakingPools[i]);
+    }
+
+    // make the current cover expire at current block
+    currentCover.period = uint32(block.timestamp) - currentCover.start;
+
+    newCover.start = uint32(block.timestamp);
+    // new period is the remaining period
+    newCover.period = currentCover.period - (uint32(block.timestamp) - currentCover.start);
+    covers.push(newCover);
+    // mint the new cover
+    _safeMint(msg.sender, newCoverId);
+
+    // the refund is proportional to the amount reduction and the period remaining
+    uint refund = currentCover.amount * amountReduction / newCover.amount * newCover.period / currentCover.period;
+
+    // adjust premium on current cover
+    currentCover.premium = currentCover.premium - uint96(refund);
+
+    uint premiumInAsset = _addPeriod(newCoverId, extraPeriod);
+    require(premiumInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+
+    if (premiumInAsset > refund) {
+      // retrieve extra required payment
+      retrievePayment(premiumInAsset- refund, newCover.payoutAsset);
+    }
+
+    // set the newly paid premium
+    newCover.premium = uint96(premiumInAsset);
+
+    return newCoverId;
+  }
+
+  function performPayoutBurn(uint coverId, address owner, uint amount) external onlyInternal override {
+    Cover memory cover = covers[coverId];
+  }
+
+
+  function retrievePayment(uint totalPrice, uint8 payoutAssetIndex) internal {
```

Comment:
`transferPremiumAsset`?

Reply:
i want to covey that the payment is being pulled from the user ( maybe `chargePremium` ? ) that was my intention here

that works too but i feel it's too general

Reply:
I like `chargePremium`